src/graph.ts

```
@@ -11,35 +11,148 @@
+        this.constructNodes = new Map<ConstructInfo, GraphNode>();
         this.cfnElementNodes = new Map<string, GraphNode>();
     }
 
     // build constructs a dependency graph from the adapter and returns its nodes sorted in topological order.
-    public static build(stack: Stack): GraphNode[] {
+    public static build(stack: StackManifest): GraphNode[] {
         const b = new GraphBuilder(stack);
         return b._build();
     }
 
+    /**
+     * Recursively parses the construct tree to create:
+     * - constructNodes
+     * - cfnElementNodes
+     *
+     * @param tree - The construct tree of the current construct being parsed
+     * @param parent - The parent construct of the construct currently being parsed
+     */
+    private parseTree(tree: ConstructTree, parent?: ConstructInfo) {
+        const construct: ConstructInfo = {
+            parent,
+            id: tree.id,
+            path: tree.path,
+            type: tree.constructInfo ? typeFromFqn(tree.constructInfo.fqn) : tree.id,
+            attributes: tree.attributes,
+        };
+        const node: GraphNode = {
+            incomingEdges: new Set<GraphNode>(),
+            outgoingEdges: new Set<GraphNode>(),
+            construct,
+        };
+        if (tree.attributes && 'aws:cdk:cloudformation:type' in tree.attributes) {
+            const cfnType = tree.attributes['aws:cdk:cloudformation:type'] as string;
+            const logicalId = this.stack.logicalIdForPath(tree.path);
+            const resource = this.stack.resourceWithLogicalId(logicalId);
+            const typ = typeFromCfn(cfnType);
+            node.construct.type = typ;
+            construct.type = typ;
+            if (resource.Type === cfnType) {
+                node.resource = resource;
+                node.logicalId = logicalId;
+                this.cfnElementNodes.set(logicalId, node);
+            } else {
+                console.error('SOMETHING WENT WRONG: ', tree);
```

I understand this is temporary, but do we want console.error or do we want throw new Error (fail fast)?

---

Updated!